README.md
# Porsche Design System - Visual Regression Tester

Export functions to create visual regression tests.

## Using the Visual Regression Tester

### Installation

Run the following command using [npm](https://www.npmjs.com):

```bash
npm install @porsche-design-system/visual-regression-tester --save-dev
```

If you prefer [Yarn](https://yarnpkg.com), use the following command instead:

```bash
yarn add @porsche-design-system/visual-regression-tester --dev
```

### Usage

__It's highly recommended executing the visual regression tester within a Docker container to get reliable tests
results across any operating system and machine.__

Check out the [Basic integration example](https://github.com/porsche-design-system/porsche-visual-regression-tester/tree/master/examples/basic-integration) for an example how to get the Visual Regression Tester up and running.

### API

#### Constructor

The constructor expects two parameters:

* `browser: Browser`
* `options: VisualRegressionTestOptions` (_optional_)

Browser should be a [Puppeteer Browser instance](https://github.com/GoogleChrome/puppeteer/blob/v1.9.0/docs/api.md#class-browser).
Check the basic integration example for [how to create a Puppeteer browser](https://github.com/porsche-design-system/porsche-visual-regression-tester/blob/master/examples/basic-integration/vrt/example-test.spec.ts#L19).

#### test() Method

In the actual visual regression test you have to call
the `test(snapshotId: string, scenario: Function, options: TestOptions = {elementSelector: '', maskSelectors: [], regressionSuffix: ''})`
-method in your expect block, taken a unique name of the reference shot as **first parameter** (`snapshotId: string`).

As **second parameter** (`scenario: Function`) within the scenario callback function you call the `goTo()` method with
the extended URL (will be concatinated with the `baseURL`), as well as `click()`, `hover()`, `focus()` and `type()` if
necessary and prepare the state to compare.
`goTo()`, `click()`, `hover()`, `focus()` and `type()` method accept following optional
parameters `networkIdleTimeout: number` and `maxInflightRequests: number` which means, — consider loading has finished
when there are no more than `maxInflightRequests` network connections for at least `networkIdleTimeout` ms.

As a **third and optional parameter** (`options: TestOptions`) you can pass following options:

* `elementSelector: string = ''` - pass a css selector for the element (selector is allowed to match exactly one element
  only) that should be included in your visual regression test.
* `maskSelectors: string[] = []` - pass a string array which includes css selectors for the elements that should be
  ignored in your visual regression test. If `maskSelectors` is used in combination with `elementSelector` then those
  two selectors are concatenated automatically to match elements nested in `elementSelector`.
* `regressionSuffix: string = ''` - pass a string to add a suffix in regression filenames

To make use of Puppeteers Page instance within the `scenario: Function` you call the `getPage()` method and apply any
supported Puppeteer method like `click()`, `hover()` or `type()`.

#### VisualRegressionTestOptions

* `viewports` selects the viewports of your browser
* `deviceScaleFactor` specify device scale factor (can be thought of as dpr)
* `fixturesDir` directory where the reference-shots will be saved
* `resultsDir` directory where the diffing- and the regression-shots will be saved
* `tolerance` gives the tolerance range for your visual regression diffs
* `baseUrl` the base URL of the page you would like to test
* `timeout` impacts the timeout limit of page load
* `mode` defines the method with which the height for snapshot is determined

*Note*: All the VisualRegressionTestOptions are optional, those are the default options:

   ```
   viewports: [320, 480, 760, 1000, 1300, 1760]
   deviceScaleFactor: 1
   fixturesDir: 'vrt/fixtures'
   resultsDir: 'vrt/results'
   tolerance: 0
   baseUrl: 'http://localhost'
   timeout: 30000
   mode: 'auto'
   ```

## Notes:

* Add your `resultsDir` folder to `.gitignore` in order to not spam git history.

## What to do when tests are failing

* Switch to your `resultsDir` directory. Here you can find the belonging `diff` and `regression` images.
* Check if you would like to accept the changes
  * **If yes**: Replace the reference shot in the `fixturesDir` folder with the belonging one in the `resultsDir` folder
    and delete the images in the `resultsDir` directory afterwards manually.
  * **If no**: Recheck your code and run the tests again, when you think you fixed it.

## License

- MIT


CHANGELOG.MD
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### [Unreleased]

### [4.5.0] - 2023-03-23

#### Removed

- retry one time in case regression occurred

### [4.4.0] - 2023-03-23

#### Added

- retry one time in case regression occurred

#### Changed

- `test()` method will fail in case no fixture exists

### [4.3.2] - 2022-06-20

### [4.3.1] - 2022-06-20

#### Changed

- npm package is prepared for public release on [npmjs.org](https://npmjs.com)

### [4.3.0] - 2021-11-29

#### Added

- Extended `TestOptions` with `initialViewportHeight`

### [4.2.1] - 2021-10-21

#### Fixed

- Exception if result image is smaller than fixture image

### [4.2.0] - 2021-10-06

#### Changed

- Improved performance by switching from Jimp to sharp

### [4.1.0] - 2021-09-16

#### Changed

- Updated npm dependencies
- Output failed viewports
- Export `TestOptions` type

### [4.0.0] - 2021-06-03

#### Changed

- Updated npm dependencies
- Upgraded `puppeteer` to `10.0.0`

### [3.0.0] - 2020-07-09

#### Changed

- Made `puppeteer` a peerDependency

### [2.0.1] - 2020-06-09

#### Changed

- Updated npm dependencies

### [2.0.0] - 2020-06-08

#### Added

- Extended `TestOptions` with `regressionSuffix`

#### Changed

- Replaced `test()` function parameters with `TestOptions` object

### [1.0.1] - 2020-05-14

- Fix support for deviceScaleFactor

### [1.0.0] - 2020-05-14

- Initial release


test/helpers/setup.ts
import 'jasmine';
import * as puppeteer from 'puppeteer';
import { Browser } from 'puppeteer';
import { VisualRegressionTester, VisualRegressionTestOptions } from '../../src/visual-regression-tester';

let browser: Browser;
let visualRegressionTester: VisualRegressionTester;
let visualRegressionRetinaTester: VisualRegressionTester;

const testOptions: VisualRegressionTestOptions = {
  viewports: [320, 480, 760, 1000, 1300, 1760],
  fixturesDir: 'test/fixtures',
  resultsDir: 'test/results',
  tolerance: 0,
  baseUrl: 'http://localhost:61423'
};

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

afterAll(async () => {
  if (browser) {
    await browser.close();
  }
});

export const getVisualRegressionTester = async (): Promise<VisualRegressionTester> => {
  if (!visualRegressionTester) {
    browser = await puppeteer.launch({
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });
    visualRegressionTester = new VisualRegressionTester(browser, testOptions);
  }

  return visualRegressionTester;
}

export const getVisualRegressionRetinaTester = async (): Promise<VisualRegressionTester> => {
  if (!visualRegressionRetinaTester) {
    browser = await puppeteer.launch({
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });
    visualRegressionRetinaTester = new VisualRegressionTester(browser, {...testOptions, viewports: [320, 350], deviceScaleFactor: 2});
  }

  return visualRegressionRetinaTester;
}


tests/helpers/setup.ts
import type { Browser } from 'puppeteer';
import type { VisualRegressionTestOptions } from '../../src/visual-regression-tester';
import puppeteer from 'puppeteer';
import { VisualRegressionTester } from '../../src/visual-regression-tester';
import { SpecReporter } from 'jasmine-spec-reporter';

let browser: Browser;
let visualRegressionTester: VisualRegressionTester;
let visualRegressionRetinaTester: VisualRegressionTester;

const testOptions: VisualRegressionTestOptions = {
  viewports: [320, 760],
  fixturesDir: 'tests/fixtures',
  resultsDir: 'tests/results',
  tolerance: 0,
  baseUrl: 'http://localhost:61423',
};

jasmine.DEFAULT_TIMEOUT_INTERVAL = 90000;

jasmine.getEnv().clearReporters();
jasmine.getEnv().addReporter(new SpecReporter());

beforeAll(async () => {
  browser = await puppeteer.launch({
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--single-process',
      '--disable-web-security',
    ],
  });
});

afterAll(async () => {
  if (browser) {
    await browser.close();
  }
});

export const getVisualRegressionTester = (): VisualRegressionTester => {
  if (!visualRegressionTester) {
    visualRegressionTester = new VisualRegressionTester(browser, testOptions);
  }

  return visualRegressionTester;
};

export const getVisualRegressionRetinaTester = (): VisualRegressionTester => {
  if (!visualRegressionRetinaTester) {
    visualRegressionRetinaTester = new VisualRegressionTester(browser, {
      ...testOptions,
      viewports: [320, 350],
      deviceScaleFactor: 2,
    });
  }

  return visualRegressionRetinaTester;
};


test/visual-regression-tester.spec.ts
import 'jasmine';
import {VisualRegressionTester} from '../src/visual-regression-tester';
import {getVisualRegressionTester, getVisualRegressionRetinaTester} from './helpers/setup';

describe('Visual-Regression-Test Tester', () => {
  let visualRegressionTester: VisualRegressionTester;

  beforeAll(async () => {
    visualRegressionTester = await getVisualRegressionTester();
  });

  it('should have no visual regression for test page', async () => {
    expect(await visualRegressionTester.test('test-page-unmasked-original', async () => {
      await visualRegressionTester.goTo('/test-page-original.html');
    })).toBeFalsy();
  });

  it('should have visual regression when compared with edited test page', async () => {
    expect(await visualRegressionTester.test('test-page-unmasked-edited', async () => {
      await visualRegressionTester.goTo('/test-page-edited.html');
    })).toBeTruthy();
  });

  it('should have no visual regression when compared with masked test page', async () => {

    expect(await visualRegressionTester.test('test-page-masked-original', async () => {
      await visualRegressionTester.goTo('/test-page-original.html');
    }, '', ['.title'])).toBeFalsy();

    expect(await visualRegressionTester.test('test-page-masked-edited', async () => {
      await visualRegressionTester.goTo('/test-page-edited.html');
    }, '', ['.title'])).toBeFalsy();
  });

  it('should have no visual regression for content within scrollable area and mask deeper nested elements', async () => {
    expect(await visualRegressionTester.test('test-page-with-scroll-area', async () => {
      await visualRegressionTester.goTo('/test-page-with-scroll-area.html');
    }, '.scrollable > .content', ['.mask'])).toBeFalsy();
  });

  it('should mask elements with an extended mask to solve issues with sub pixel rendering', async () => {
    expect(await visualRegressionTester.test('test-page-extend-mask-bounding', async () => {
      await visualRegressionTester.goTo('/test-page-extend-mask-bounding.html');
    }, '', ['.mask'])).toBeFalsy();
  });

  it('should ignore mask selectors that are invisible without size definition', async () => {
    expect(await visualRegressionTester.test('test-page-ignores-invisible-mask-selectors', async () => {
      await visualRegressionTester.goTo('/test-page-ignores-invisible-mask-selectors.html');
    }, '', ['.mask'])).toBeFalsy();
  });

  it('should properly overlays with fixed position', async () => {
    expect(await visualRegressionTester.test('test-page-with-fixed-positioning', async () => {
      await visualRegressionTester.goTo('/test-page-with-fixed-positioning.html');
    }, '', ['.modal'])).toBeFalsy();
  });

  describe('interaction', () => {

    it('should wait for page interaction "click" before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-click', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction.html');
        await visualRegressionTester.click('button');
      })).toBeFalsy();
    });

    it('should wait for page interaction "hover" before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-hover', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction.html');
        await visualRegressionTester.hover('button');
      })).toBeFalsy();
    });

    it('should wait for page interaction "focus" before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-focus', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction.html');
        await visualRegressionTester.focus('button');
      })).toBeFalsy();
    });

    it('should wait for page interaction "type" before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-type', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction.html');
        await visualRegressionTester.type('input', 'car');
      })).toBeFalsy();
    });
  });

  describe('network idle', () => {

    it('should wait for no more than 0 network connections for at least 3 seconds', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-zero-network-connections', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-zero-network-connections.html', 3000);
      })).toBeFalsy();
    });

    it('should wait for no more than 0 network connections for at least 0.5 seconds and ignore possibly delayed network connections (e.g. polling every 2 seconds)', async () => {
      expect(await visualRegressionTester.test('/test-page-waits-for-zero-network-connections-and-ignores-delayed-ones', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-zero-network-connections.html', 500);
      })).toBeFalsy();
    });

    it('should wait for page interaction "click" and network requests (polling) before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-and-network-requests-click', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction-and-network-requests.html');
        await visualRegressionTester.click('button', 2000);
      }, '', ['button'])).toBeFalsy();
    });

    it('should wait for page interaction "hover" and network requests (polling) before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-and-network-requests-hover', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction-and-network-requests.html');
        await visualRegressionTester.hover('button', 2000);
        await visualRegressionTester.getPage().waitFor(100);
      }, '', ['button'])).toBeFalsy();
    });

    it('should wait for page interaction "focus" and network requests (polling) before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-and-network-requests-focus', async () => {
        await visualRegressionTester.goTo('/test-page-waits-for-user-interaction-and-network-requests.html');
        await visualRegressionTester.focus('button', 2000);
      }, '', ['button'])).toBeFalsy();
    });

    it('should wait for page interaction "type" and network requests (polling) before comparing snapshots', async () => {
      expect(await visualRegressionTester.test('test-page-waits-for-user-interaction-and-network-requests-type',
        async () => {
          await visualRegressionTester.goTo('/test-page-waits-for-user-interaction-and-network-requests.html');
          await visualRegressionTester.type('input', 'c', 2000);
        })).toBeFalsy();
    });
  });
});


describe('Visual-Regression-Test Retina Tester', () => {
  let vrt: VisualRegressionTester;

  beforeAll(async () => {
    vrt = await getVisualRegressionRetinaTester();
  });

  it('should have no visual regression for test page', async () => {
    expect(await vrt.test('test-page-unmasked-original-retina', async () => {
      await vrt.goTo('/test-page-original.html');
    })).toBeFalsy();
  });
});


src/visual-regression-tester.ts
import * as fs from 'fs';
import * as path from 'path';
import { Browser, ClickOptions, ElementHandle, Page, PuppeteerLifeCycleEvent } from 'puppeteer';
import sharp from 'sharp';
import pixelmatch from 'pixelmatch';

export type VisualRegressionTestOptions = {
  viewports?: number[];
  deviceScaleFactor?: number;
  fixturesDir?: string;
  resultsDir?: string;
  tolerance?: number;
  baseUrl?: string;
  timeout?: number;
  mode?: 'auto' | 'square-auto';
  waitUntilMethod?: PuppeteerLifeCycleEvent;
};

export type TestOptions = {
  elementSelector?: string;
  maskSelectors?: string[];
  regressionSuffix?: string;
  initialViewportHeight?: number;
};

type Writeable<T> = { -readonly [P in keyof T]: T[P] };

type WritableDomRect = Writeable<Omit<DOMRect, 'toJson'>>;

export class VisualRegressionTester {
  private readonly options: VisualRegressionTestOptions = {
    viewports: [320, 480, 760, 1000, 1300, 1760],
    deviceScaleFactor: 1,
    fixturesDir: 'vrt/fixtures',
    resultsDir: 'vrt/results',
    tolerance: 0,
    baseUrl: 'http://localhost',
    timeout: 30000,
    mode: 'auto',
    waitUntilMethod: 'networkidle0',
  };
  private page: Page;

  constructor(private browser: Browser, options: VisualRegressionTestOptions = {}) {
    this.options = {
      ...this.options,
      ...options,
    };
  }
  getPage(): Page {
    return this.page;
  }

  async goTo(url: string, networkIdleTimeout: number = 500, maxInflightRequests: number = 0): Promise<void> {
    await Promise.all([
      this.waitForNetworkIdle(networkIdleTimeout, maxInflightRequests),
      this.page.goto(`${this.options.baseUrl}${url}`, { waitUntil: this.options.waitUntilMethod }),
    ]);
  }

  async click(
    selector: string,
    networkIdleTimeout: number = 500,
    maxInflightRequests: number = 0,
    options?: Partial<ClickOptions>
  ): Promise<void> {
    await Promise.all([
      this.waitForNetworkIdle(networkIdleTimeout, maxInflightRequests),
      this.page.click(selector, options),
    ]);
  }

  async focus(selector: string, networkIdleTimeout: number = 500, maxInflightRequests: number = 0): Promise<void> {
    await Promise.all([this.waitForNetworkIdle(networkIdleTimeout, maxInflightRequests), this.page.focus(selector)]);
  }

  async type(
    selector: string,
    input: string,
    networkIdleTimeout: number = 500,
    maxInflightRequests: number = 0
  ): Promise<void> {
    await Promise.all([
      this.waitForNetworkIdle(networkIdleTimeout, maxInflightRequests),
      this.page.type(selector, input),
    ]);
  }

  async hover(selector: string, networkIdleTimeout: number = 500, maxInflightRequests: number = 0): Promise<void> {
    await Promise.all([this.waitForNetworkIdle(networkIdleTimeout, maxInflightRequests), this.page.hover(selector)]);
  }

  async test(snapshotId: string, scenario: Function, options?: TestOptions): Promise<boolean> {
    sharp.cache(false);

    const opts: TestOptions = {
      elementSelector: '',
      maskSelectors: [],
      regressionSuffix: '',
      initialViewportHeight: 1,
      ...options,
    };

    const errors: number[] = [];

    for (const viewport of this.options.viewports) {
      const { regressionSuffix } = opts;
      const { fixturesDir, resultsDir } = this.options;
      const suffix = regressionSuffix ? '.' + regressionSuffix : '';

      const paths = {
        reference: `${fixturesDir}/${snapshotId}.${viewport}.png`,
        regression: `${resultsDir}/${snapshotId}${suffix}.${viewport}.png`,
        diff: `${resultsDir}/${snapshotId}${suffix}.${viewport}.diff.png`,
      };

      this.page = await this.newPage(viewport, opts.initialViewportHeight);

      this.cleanSnapshots([paths.regression, paths.diff]);

      await scenario();

      const height = await this.page.evaluate(() => document.body.clientHeight);
      await this.page.setViewport({
        width: viewport,
        height: height,
        deviceScaleFactor: this.options.deviceScaleFactor,
      });

      if (fs.existsSync(paths.reference)) {
        const fixture = sharp(path.resolve(paths.reference));
        let regression = await this.compareSnapshots(fixture, opts.elementSelector, opts.maskSelectors);

        // TODO: make it configurable by `retries` option
        if (regression) {
          console.log('Regression occured, retry for viewport:', viewport);
          regression = await this.compareSnapshots(fixture, opts.elementSelector, opts.maskSelectors); // try one more time
        }

        if (regression) {
          fs.mkdirSync(path.resolve(this.options.resultsDir), { recursive: true });

          errors.push(viewport);
          await regression.result.toFile(paths.regression);
          await regression.diff.toFile(paths.diff);
        }
      } else {
        errors.push(viewport); // test execution shall fail, in case no fixture exists
        const reference = await this.createSnapshot(opts.elementSelector, opts.maskSelectors);
        await reference.toFile(paths.reference);
      }

      await this.page.close();
    }

    if (this.options.viewports.length > 1 && errors.length) {
      console.log('Failed viewports:', errors.join(', '));
    }

    return errors.length > 0;
  }

  public waitForNetworkIdle(timeout: number = 500, maxInflightRequests: number = 0): Promise<void> {
    const onRequestStarted = () => {
      ++inflight;
      if (inflight > maxInflightRequests) {
        clearTimeout(timeoutId);
      }
    };

    const onRequestFinished = () => {
      if (inflight === 0) {
        return;
      }
      --inflight;
      if (inflight === maxInflightRequests) {
        timeoutId = setTimeout(onTimeoutDone, timeout);
      }
    };

    const onTimeoutDone = () => {
      this.page.off('request', onRequestStarted);
      this.page.off('requestfinished', onRequestFinished);
      this.page.off('requestfailed', onRequestFinished);
      fulfill();
    };

    let inflight = 0;
    let fulfill: (value?: void | PromiseLike<void>) => void;
    let timeoutId = setTimeout(onTimeoutDone, timeout);

    this.page.on('request', onRequestStarted);
    this.page.on('requestfinished', onRequestFinished);
    this.page.on('requestfailed', onRequestFinished);

    return new Promise((x) => (fulfill = x));
  }

  private async newPage(viewport: number, viewportHeight: number): Promise<Page> {
    const page = await this.browser.newPage();
    page.setDefaultNavigationTimeout(this.options.timeout);
    await page.setViewport({
      width: viewport,
      height: this.options.mode === 'square-auto' ? viewport : viewportHeight,
      deviceScaleFactor: this.options.deviceScaleFactor,
    });

    return page;
  }

  private async createSnapshot(elementSelector: string, maskSelectors: string[]): Promise<sharp.Sharp> {
    const buffer = await (elementSelector
      ? ((
          await this.page.$(elementSelector)
        ).screenshot({
          captureBeyondViewport: false,
        }) as unknown as Promise<string>)
      : (this.page.screenshot({
          fullPage: true,
          captureBeyondViewport: false,
        }) as unknown as Promise<string>));

    const rawImage = sharp(buffer);

    return maskSelectors.length ? await this.maskSnapshot(rawImage, elementSelector, maskSelectors) : rawImage;
  }

  private async maskSnapshot(
    image: sharp.Sharp,
    elementSelector: string,
    maskSelectors: string[]
  ): Promise<sharp.Sharp> {
    // BoundingBoxes of Mask elements
    const boundingBoxes: WritableDomRect[] = [];

    for (const maskSelector of maskSelectors) {
      const maskElements = await this.page.$$(`${elementSelector} ${maskSelector}`);

      for (const maskElement of maskElements) {
        const boundingBox = await this.getBoundingBox(maskElement);
        if (boundingBox) {
          boundingBoxes.push(boundingBox);
        }
      }
    }

    const { width: imageWidth, height: imageHeight } = await image.metadata();
    // BoundingBox of Element to screenshot
    const elementBoundingBox: WritableDomRect = elementSelector
      ? await this.getBoundingBox(await this.page.$(elementSelector), 0)
      : ({
          left: 0,
          right: imageWidth,
          top: 0,
          height: imageHeight,
          width: imageWidth,
          bottom: imageHeight,
          x: 0,
          y: 0,
        } as WritableDomRect);

    image.composite(
      boundingBoxes.map((maskBoundingBox) => {
        // Adjust the size of the mask to fit in the area of the screenshot
        // ┌────────────┐
        // │ ┌────────┐ │
        // │ │  ┌─────┼─┤
        // │ │  │Mask │┼│
        // │ │  └─────┼─┤
        // │ └────────┘ │
        // └────────────┘

        if (maskBoundingBox.right > elementBoundingBox.right) {
          maskBoundingBox.width = maskBoundingBox.width - (maskBoundingBox.right - elementBoundingBox.right);
        }

        // ┌────────────┐
        // │ ┌────────┐ │
        // │ │ ┌────┐ │ │
        // │ │ │Mask│ │ │
        // │ └─┼────┼─┘ │
        // │   │┼┼┼┼│   │
        // └───┴────┴───┘

        if (maskBoundingBox.bottom > elementBoundingBox.bottom) {
          maskBoundingBox.height = maskBoundingBox.height - (maskBoundingBox.bottom - elementBoundingBox.bottom);
        }

        // ┌────────────┐
        // │ ┌────────┐ │
        // ├─┼─────┐  │ │
        // │┼│Mask │  │ │
        // ├─┼─────┘  │ │
        // │ └────────┘ │
        // └────────────┘

        if (maskBoundingBox.left < elementBoundingBox.left) {
          maskBoundingBox.width = maskBoundingBox.width - (elementBoundingBox.left - maskBoundingBox.left);
          maskBoundingBox.left = 0;
        }

        // ┌───┬────┬───┐
        // │   │┼┼┼┼│   │
        // │ ┌─┼────┼─┐ │
        // │ │ │Mask│ │ │
        // │ │ └────┘ │ │
        // │ └────────┘ │
        // └────────────┘

        if (maskBoundingBox.top < elementBoundingBox.top) {
          maskBoundingBox.height = maskBoundingBox.height - (elementBoundingBox.top - maskBoundingBox.top);
          maskBoundingBox.top = 0;
        }

        return {
          input: {
            create: {
              width: maskBoundingBox.width,
              height: maskBoundingBox.height,
              channels: 3,
              background: '#FF00FF',
            },
          },
          left: maskBoundingBox.left,
          top: maskBoundingBox.top,
        };
      })
    );

    return image;
  }

  private async getBoundingBox(element: ElementHandle, extendOuterBounds: number = 1): Promise<WritableDomRect> {
    const boundingBox = await element.evaluate((el) => {
      const { width, height, left, right, top, bottom } = el.getBoundingClientRect();
      return { width, height, left, right, top, bottom } as WritableDomRect;
    });

    if (boundingBox !== null && boundingBox.width !== 0 && boundingBox.height !== 0) {
      return {
        ...boundingBox,
        width: Math.ceil(boundingBox.width + extendOuterBounds * 2),
        height: Math.ceil(boundingBox.height + extendOuterBounds * 2),
        left: Math.floor(boundingBox.left - extendOuterBounds),
        right: Math.ceil(boundingBox.right + extendOuterBounds),
        top: Math.floor(boundingBox.top - extendOuterBounds),
        bottom: Math.ceil(boundingBox.bottom + extendOuterBounds),
      };
    }
  }

  private async compareSnapshots(
    fixture: sharp.Sharp,
    elementSelector: string,
    maskSelectors: string[]
  ): Promise<{ result: sharp.Sharp; diff: sharp.Sharp }> {
    const result = await this.createSnapshot(elementSelector, maskSelectors);
    const resultClone = result.clone();
    let { info: resultInfo, data: resultData } = await result.raw().toBuffer({ resolveWithObject: true });
    const { info: fixtureInfo, data: fixtureData } = await fixture.raw().toBuffer({ resolveWithObject: true });

    if (resultData.compare(fixtureData) !== 0) {
      let { height: resultHeight, width: resultWidth } = resultInfo;
      const { height: fixtureHeight } = fixtureInfo;

      if (resultHeight > fixtureHeight) {
        fixture.extend({
          bottom: resultHeight - fixtureHeight,
        });
      } else if (resultHeight < fixtureHeight) {
        result.extend({
          bottom: fixtureHeight - resultHeight,
        });

        const resultInfoData = await result.raw().toBuffer({ resolveWithObject: true });
        resultInfo = resultInfoData.info;
        resultData = resultInfoData.data;
        resultHeight = resultInfo.height;
      }

      const diff = resultData;
      const fixtureRaw = await fixture.toBuffer();

      const pixelDiff = pixelmatch(resultData, fixtureRaw, diff, resultWidth, resultHeight, {
        threshold: this.options.tolerance,
      });

      // To avoid breaking changes we need to calculate the threshold like it was with Jimp
      const percent = pixelDiff / (resultWidth * resultHeight);

      if (percent !== 0) {
        return {
          result: resultClone,
          diff: sharp(diff, { raw: resultInfo }),
        };
      }
    }
  }

  private cleanSnapshots(paths: string[]): void {
    paths
      .map((p) => path.resolve(p))
      .forEach((file) => {
        if (fs.existsSync(file)) {
          fs.unlinkSync(file);
        }
      });
  }
}
